src/java.base/share/classes/com/sun/crypto/provider/PBMAC1Parameters.java

```diff
@@ -0,0 +1,244 @@
+
+    private static final ObjectIdentifier pkcs5PBKDF2_OID =
+            ObjectIdentifier.of(KnownOIDs.PBKDF2WithHmacSHA1);
+
+    // the PBMAC1 algorithm name
+    private String pbmac1AlgorithmName = null;
+
+    private byte[] salt = null;
+
+    // Iteration count
+    private int iCount = 0;
+
+    // the key derivation function (default is HmacSHA1)
+    private final ObjectIdentifier kdfAlgo_OID =
+            ObjectIdentifier.of(KnownOIDs.HmacSHA1);
+
+    // length of key generated by the kdf
+    private int keyLength = -1;
+
+    protected void engineInit(AlgorithmParameterSpec paramSpec)
+        throws InvalidParameterSpecException
+    {
+        if (!(paramSpec instanceof PBEParameterSpec)) {
+            throw new InvalidParameterSpecException
+                    ("Inappropriate parameter specification");
+        }
+        salt = ((PBEParameterSpec)paramSpec).getSalt().clone();
+        iCount = ((PBEParameterSpec)paramSpec).getIterationCount();
+    }
+
+    protected void engineInit(byte[] encoded)
+        throws IOException
+    {
+        DerValue pBMAC1_params = new DerValue(encoded);
+        if (pBMAC1_params.tag != DerValue.tag_Sequence) {
+            throw new IOException("PBMAC1 parameter parsing error: "
+                    + "not an ASN.1 SEQUENCE tag");
+        }
+        DerValue[] Info = (new DerInputStream(pBMAC1_params.toByteArray()))
+                .getSequence(2);
+        if (Info.length != 2) {
+            throw new IOException("PBMAC1 parameter parsing error: "
+                + "expected length not 2");
+        }
+        ObjectIdentifier OID = Info[1].data.getOID();
+            KnownOIDs o = KnownOIDs.findMatch(OID.toString());
+            if (o == null || (!o.stdName().equals("HmacSHA1") &&
+                    !o.stdName().equals("HmacSHA224") &&
+                    !o.stdName().equals("HmacSHA256") &&
+                    !o.stdName().equals("HmacSHA384") &&
+                    !o.stdName().equals("HmacSHA512") &&
+                    !o.stdName().equals("HmacSHA512/224") &&
+                    !o.stdName().equals("HmacSHA512/256"))) {
+                throw new IOException("PBMAC1 parameter parsing error: "
+                        + "expecting the object identifier for a HmacSHA key "
+                        + "derivation function");
+            }
+        // Hmac function used to compute the MAC
+        String hmacAlgo = o.stdName();
+
+        DerValue kdf = pBMAC1_params.data.getDerValue();
+        var kdfParams = new PBKDF2Parameters();
+        String kdfAlgo = kdfParams.init(kdf);
+        salt = kdfParams.getSalt();
+        iCount = kdfParams.getIterationCount();
+
+        // Key length must be present. It is not currently used.
+        keyLength = kdfParams.getKeyLength();
+        if (keyLength == -1) {
+            throw new IOException("PBMAC1 parameter parsing "
+                    + "error: missing keyLength field");
+        }
```

Comment:
Isn't this a PKCS#12 restriction? If so, this should be moved to PKCS12KeyStore class?

Reply:
The `keyLength` is present check should stay for now since it is a MUST. We can revisit this later if we switch to a different parameter spec. Without a `PBMAC1ParameterSpec` there is no way to pass `keyLength` to PKCS12.